less_users/views.py

```diff
@@ -75,69 +84,112 @@ def get_queryset(self):
     user_id = request.user
-    if 'done' in request.POST:
-        """should check if points can be added (depending on durration and frequncy in model"""
-        challenge_id = int(request.POST.get('done'))
-        points = Challenge.objects.get(id=challenge_id).points
-        this = UserChallenge.objects.get(user_id=user_id, challenge_id=challenge_id, is_active=True)
-        user_challenge_id = this.id
-        this.check_if_active()
-        if this.get_points == points and this.is_active:
-            new_log = Log.objects.create(user_challenge_id=user_challenge_id, points=points)
-            my_profile = Profile.objects.get(user_id=user_id)
-            my_profile.points += points
-            my_profile.save()
-            messages.success(request, f'You have got new points {points} for challenge: {this.challenge.name}!')
-        elif this.get_points == 0 or this.is_active == False:
-            messages.warning(request, f'You cannot get new points for this challenge yet: {this.challenge.name}!')
-    elif 'stop' in request.POST:
-        user_challenge_id = int(request.POST.get('stop'))
-        this = UserChallenge.objects.get(id=user_challenge_id)
-        this.is_active = False
-        this.save()
-        messages.info(request, f'You stopped challenge: {this.challenge.name}. You can activate the new one.')
-    elif 'delete' in request.POST:
-        user_challenge_id = int(request.POST.get('delete'))
-        this = UserChallenge.objects.get(id=user_challenge_id)
-        this.delete()
-        messages.warning(request, f'You have deleted {this.challenge.name} from your challenges')
-    return redirect('my_challenges')
-
-
-
+    challenge_id = int(request.POST.get("done"))
+    points = Challenge.objects.get(id=challenge_id).points
+    user_challenge = UserChallenge.objects.get(
+        user_id=user_id, challenge_id=challenge_id, is_active=True
+    )
+    user_challenge.check_if_active()
+    if user_challenge.get_points == points and user_challenge.is_active:
+        new_log = Log.objects.create(user_challenge_id=user_challenge.id, points=points)
+        my_profile = Profile.objects.get(user_id=user_id)
+        my_profile.points += points
+        my_profile.save()
+        messages.success(
+            request,
+            f"You have just got {points} points for challenge: {user_challenge.challenge}!",
+        )
+    elif user_challenge.get_points == 0 or not user_challenge.is_active:
+        messages.warning(
+            request,
+            f"You cannot get new points for this challenge yet: {user_challenge.challenge}. "
+            f"You have already did it in last {user_challenge.challenge.frequency} days "
+            f"or its duration period passed",
+        )
+    return redirect("my_challenges")
+
+
+def event_view_stop(request):
+    """stop user challenge despite its duration"""
+    user_challenge_id = int(request.POST.get("stop"))
```

Comment:
co jeżeli stop nie jest konwertowalne na inta?

Reply:
narazie testuję wersję nie zmienioną, gdy posługuję się id-kami... masz rację, że jak zmienię to na stringify, to będzie klops. Ale jeszcze działa, a request bierze int z html-a.

Reply:
Teraz też jest klops. Zobacz co się stanie jeżeli wyślesz request z wartością stop ustawioną na „klops” :P

Reply:
Zrobiłam narazie w jednym z 3 podobnych widoków _def event_view_done(request)_ **łapanie wyjątku ValueError** przez try/except (gdy zamiast int podany jest np. string) oraz **if statement** dla przypadku, gdyby ktoś podał int'a spoza zakresu bazy danych, i nie było takiego rekordu (exists()). Pojawiają się wtedy odpowiednie error.messages. Czy to o to Ci chodziło, żeby zabezpieczyć? Tylko strasznie mi się ta funkcja rozrosła

Reply:
tak, chodziło mi o to, żeby łapać wszystkie potencjalne wyjątki. Ale musisz to dorzucić tak naprawdę w każdym miejscu. Możesz to dorzucić w kolejnych widokach w kolejnej prce, ale to jest must have.
Poza tym funkcja ci się rozrosła a i tak nie wyłapałaś wszystkiego co należy wyłapać ;)